src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+   * @param  shortcutid  Graph Id of the shortcut edge.
+   * @return Returns the edgeids of the directed edges this shortcut represents.
+   */
+  std::vector<valhalla::baldr::GraphId>
+  recover_shortcut(valhalla::baldr::GraphReader& reader,
+                   const valhalla::baldr::GraphId& shortcut_id) const {
+    using namespace valhalla::baldr;
+    // grab the shortcut edge
+    const GraphTile* tile = reader.GetGraphTile(shortcut_id);
+    const DirectedEdge* shortcut = tile->directededge(shortcut_id);
+
+    // bail if this isnt a shortcut
+    if (!shortcut->is_shortcut()) {
+      return {shortcut_id};
+    }
+
+    // loop over the edges leaving its begin node and find the superseded edge
+    GraphId begin_node = reader.edge_startnode(shortcut_id);
+    if (!begin_node)
+      return {shortcut_id};
+
+    // loop over the edges leaving its begin node and find the superseded edge
+    std::vector<GraphId> edges;
+    for (const DirectedEdge& de : tile->GetDirectedEdges(begin_node.id())) {
+      if (shortcut->shortcut() & de.superseded()) {
+        edges.push_back(tile->header()->graphid());
+        edges.back().set_id(&de - tile->directededge(0));
+        break;
+      }
+    }
+
+    // bail if we couldnt find it
+    if (edges.empty()) {
+      LOG_TRACE("Unable to recover shortcut for edgeid " + std::to_string(shortcut_id) +
+                " | no superseded edge");
+      return {shortcut_id};
+    }
+
+    // seed the edge walking with the first edge
+    const DirectedEdge* current_edge = tile->directededge(edges.back());
+    uint32_t accumulated_length = current_edge->length();
+
+    // walk edges until we find the same ending node as the shortcut
+    while (current_edge->endnode() != shortcut->endnode()) {
+      // get the node at the end of the last edge we added
+      const NodeInfo* node = reader.GetEndNode(current_edge, tile);
+      if (!node)
+        return {shortcut_id};
+      auto node_index = node - tile->node(0);
+
+      // check the edges leaving this node to see if we can find the one that is part of the shortcut
+      current_edge = nullptr;
+      for (const DirectedEdge& edge : tile->GetDirectedEdges(node_index)) {
+        // are they the same enough that its part of the shortcut
+        // NOTE: this fails in about .05% of cases where there are two candidates and its not clear
+        // which edge is the right one. looking at shortcut builder its not obvious how this is
+        // possible as it seems to terminate a shortcut if more than one edge pair can be
+        // contracted... NOTE: because we change the speed of the edge in graph enhancer we cant use
+        // speed as a reliable determining factor
+        if (begin_node != edge.endnode() && !edge.is_shortcut() &&
+            edge.forwardaccess() == shortcut->forwardaccess() &&
+            edge.reverseaccess() == shortcut->reverseaccess() && edge.sign() == shortcut->sign() &&
```

Comment:
previously we only recovered shortcuts on edges with forward auto access. ive changed that to just say that access is the same in both directions of the edge, whcih allows us to recover opposing shortcuts of oneways. in a recent planet dataset the stats were 

```
2020/12/07 21:23:37.570000 [INFO] Shortcut recovery cache enabled
2020/12/07 21:23:47.595293 [INFO] 14191294 shortcuts recovered as 90807070 superseded edges. 1776 shortcuts could not be recovered.
```

Reply:
also notice recovery only takes 10 seconds, hence why we default to turning it off and doing on the fly recovery. caching of this costs approximately 1gig of ram per process. hardly worth the tiny amount of cpu it takes to recover on the fly